File: 20221014_Friday/20221014.go
// Leetcode: 2095. Delete the Middle Node of a Linked List.
// https://leetcode.com/problems/delete-the-middle-node-of-a-linked-list/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 342 ms, faster than 87.36% of Go online submissions for Delete the Middle Node of a Linked List.
// Memory Usage: 12.1 MB, less than 28.74% of Go online submissions for Delete the Middle Node of a Linked List.
// 2022.10.14 Daily Callenge.
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteMiddle(head *ListNode) *ListNode {
  count := 0
  tmp := head
  for ; nil != tmp ; {
    count += 1
    tmp = tmp.Next
  }
  if 1 == count {
    return nil
  }
  if 2 == count {
    head.Next = nil
    return head
  }
  place := (count/2) - 1
  tmp = head
  for i := 0; i <= place; i += 1{
    if i == place {
      tmp.Next = tmp.Next.Next
    } else {
      tmp = tmp.Next
    }
  }
  return head
}

File: 20221014_Friday/20221014.rb
# Leetcode: 2095. Delete the Middle Node of a Linked List.
# https://leetcode.com/problems/delete-the-middle-node-of-a-linked-list/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 2340 ms, faster than 22.22% of Ruby online submissions for Delete the Middle Node of a Linked List.
# Memory Usage: 234.5 MB, less than 44.44% of Ruby online submissions for Delete the Middle Node of a Linked List.
# 2022.10.14 Daily Challenge.
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def delete_middle(head)
  arr = []
  while head 
    arr.push(head)
    head = head.next
  end
  arr.each{|x| x.next = nil}
  return nil if 1 == arr.size
  prev = nil
  skip = arr.size/2  
  (0...arr.size).each do |i|
    next if i == skip
    prev.next = arr[i] if prev
    prev = arr[i]
  end
  return arr[0]
end
